src/interp.rs

```diff
@@ -55,97 +59,160 @@ pub struct TypedVal<'tcx> {
-/// during execution.
-fn eval_terminator<'a,'tcx>(terminator: &Terminator<'tcx>,
-                            vm: &mut Machine<'a, 'tcx>) {
-    let Terminator {
-        ref kind,
-        ..
-    } = *terminator;
+    /// A `Place` is a MIR construct which describes a pointer into some position in
+    /// memory (be it local, static, or on the heap). In Mirkat, we use the
+    /// information from the `Place` to create an `Address` - an enum pointing to a
+    /// location inside the Machine.
+    fn eval_place(&mut self, place: &Place<'tcx>) -> Address {
+        match *place {
+            Place::Local(local) => Address::Local(local),
+            Place::Static(ref static_) => unimplemented!(),
+            Place::Projection(ref proj) => unimplemented!(),
+        }
+    }
 
-    match *kind {
-        TerminatorKind::Goto { target } => {
-            eval_basic_block(target, vm)
+    /// An operand's value can always be determined without needing to temporarily
+    /// save values to a stack. This is because of the recursive nature of the MIR.
+    /// An operand will only ever return a primitive value or a pointer.
+    fn eval_operand(&mut self,
+                    operand: &Operand<'tcx>)
+                    -> TypedVal<'tcx> {
+        let ty = operand.ty(self.cur_frame().mir, self.tcx);
+        match *operand {
+            Operand::Copy(ref place) |
+            Operand::Move(ref place) => {
+                unimplemented!()
+            }
+            Operand::Constant(ref constant) => {
+                let Constant {ref literal, ..} = **constant;
+                let val = match *literal {
+                    Literal::Value { ref value } => value.val.to_bytes(),
+                    Literal::Promoted { index } => {
+                        unimplemented!()
+                    }
+                };
+                return TypedVal {
+                    val,
+                    ty
+                }
+            }
         }
-        _ => unimplemented!()
     }
 }
 
-/// A `Place` is a MIR construct which describes a pointer into some position in
-/// memory (be it local, static, or on the heap). In Mirkat, we use the
-/// information from the `Place` to create an `Address` - a struct pointing to a
-/// location inside the Machine.
-fn eval_place<'tcx>(place: &Place<'tcx>) -> Address {
-    match *place {
-        Place::Local(local) => Address::Local(local),
-        Place::Static(ref static_) => unimplemented!(),
-        Place::Projection(ref proj) => unimplemented!(),
-    }
+/// Builds a VM and starts interpretation from the given MIR function, this is
+/// normally the main method.
+pub fn entry_point<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                             def_id: DefId) {
+    let m = machine::Memory::new(MEMORY_CAPACITY);
+    let mut vm = Machine::new(m, tcx);
+    vm.eval_fn_call(def_id, None);
 }
 
+fn exit(status_code: i32) {
```

Comment:
I'm not sure providing a wrapper around `process::exit` is useful: why not call the function directly?

Reply:
Agreed

Reply:
The thinking was so that if I wanted in the future to provide debugging information on exit, I had a single place to update. But I see what you mean, it doesn't feel very useful really.